File: services/cart/src/main/java/com/commerceos/cart/domain/CartItemEntity.java
package com.commerceos.cart.domain;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "cart_items", indexes = {
    @Index(name = "idx_cart_customer_sku", columnList = "customerId,sku", unique = true)
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CartItemEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 64)
    private String customerId;

    private String medicineId;

    @Column(nullable = false, length = 64)
    private String sku;

    @Column(nullable = false, length = 255)
    private String name;

    private String brand;

    private String packSize;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal unitPrice;

    @Column(precision = 12, scale = 2)
    private BigDecimal mrp;

    @Column(nullable = false)
    private Integer quantity;

    @Column(nullable = false)
    private Boolean prescriptionRequired;

    @Column(nullable = false)
    private Boolean coldChain;

    private String image;

    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    @Column(nullable = false)
    private Instant updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = Instant.now();
        this.updatedAt = Instant.now();
        if (this.quantity == null) this.quantity = 1;
        if (this.prescriptionRequired == null) this.prescriptionRequired = false;
        if (this.coldChain == null) this.coldChain = false;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = Instant.now();
    }
}

File: apps/android/app/src/main/java/com/commerceos/android/data/local/CartItemEntity.kt
package com.commerceos.android.data.local

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.commerceos.android.model.CartItem
import com.commerceos.android.model.PharmacyCartAttributes
import java.math.BigDecimal

@Entity(tableName = "cart_items")
data class CartItemEntity(
    @PrimaryKey
    val sku: String,
    val customerId: String,
    val productId: String,
    val name: String,
    val unitPrice: Double,
    val quantity: Int,
    val verticalId: String = "general",
    val merchantId: String? = null,
    val mrp: Double = unitPrice,
    val brand: String? = null,
    val packSize: String? = null,
    val image: String? = null,
    val prescriptionRequired: Boolean = false,
    val coldChain: Boolean = false,
    val updatedAt: Long = System.currentTimeMillis()
) {
    fun toCartItem(): CartItem = CartItem(
        productId = productId.ifBlank { sku },
        sku = sku,
        name = name.ifBlank { "Medicine Item" },
        unitPrice = BigDecimal.valueOf(unitPrice),
        quantity = quantity.coerceAtLeast(1),
        verticalId = verticalId.ifBlank { "general" },
        merchantId = merchantId,
        mrp = BigDecimal.valueOf(if (mrp > 0.0) mrp else unitPrice),
        brand = brand,
        packSize = packSize,
        image = image,
        pharmacyAttributes = if (prescriptionRequired || coldChain) {
            PharmacyCartAttributes(
                prescriptionRequired = prescriptionRequired,
                coldChain = coldChain
            )
        } else null
    )

    companion object {
        fun fromCartItem(item: CartItem, customerId: String): CartItemEntity = CartItemEntity(
            sku = item.sku,
            customerId = customerId,
            productId = (item.productId ?: item.sku).ifBlank { item.sku },
            name = (item.name ?: "Medicine Item").ifBlank { "Medicine Item" },
            unitPrice = (item.unitPrice ?: BigDecimal.ZERO).toDouble(),
            quantity = (item.quantity ?: 1).coerceAtLeast(1),
            verticalId = (item.verticalId ?: "general").ifBlank { "general" },
            merchantId = item.merchantId,
            mrp = (item.mrp ?: item.unitPrice ?: BigDecimal.ZERO).toDouble(),
            brand = item.brand,
            packSize = item.packSize,
            image = item.image,
            prescriptionRequired = item.prescriptionRequired,
            coldChain = item.coldChain,
            updatedAt = System.currentTimeMillis()
        )
    }
}
